source/opt/ir_builder.h

```diff
@@ -0,0 +1,144 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef LIBSPIRV_OPT_IR_BUILDER_H_
+#define LIBSPIRV_OPT_IR_BUILDER_H_
+
+#include "opt/basic_block.h"
+#include "opt/instruction.h"
+#include "opt/ir_context.h"
+
+namespace spvtools {
+namespace opt {
+
+// In SPIR-V, ids are encoded as uint16_t, this id is guarantied to be always
+// invalid.
+constexpr uint32_t kInvalidId = (uint32_t)-1;
+
+// Helper class to abstract instruction construction and insertion.
+// |AnalysisToPreserve| asks the InstructionBuilder to preserve requested
+// analysis.
+// Supported analysis:
+//   - Def-use analysis
+template <ir::IRContext::Analysis AnalysisToPreserve =
+              ir::IRContext::kAnalysisNone>
+class InstructionBuilder {
+  static_assert(!(AnalysisToPreserve & ~ir::IRContext::kAnalysisDefUse),
+                "Only Def-use analysis update is supported");
+
+ public:
+  using InsertionPointTy = spvtools::ir::BasicBlock::iterator;
+
+  InstructionBuilder(ir::IRContext* context, InsertionPointTy insert_before)
+      : context_(context), insert_before_(insert_before) {}
+
+  // Creates a new selection merge instruction.
+  // The id |merge_id| is the merge basic block id.
+  ir::Instruction* AddSelectionMerge(uint32_t merge_id) {
+    std::unique_ptr<ir::Instruction> new_branch_merge(new ir::Instruction(
+        GetContext(), SpvOpSelectionMerge, 0, 0,
+        {{spv_operand_type_t::SPV_OPERAND_TYPE_ID, {merge_id}},
+         {spv_operand_type_t::SPV_OPERAND_TYPE_LITERAL_INTEGER, {0}}}));
+    return AddInstruction(std::move(new_branch_merge));
+  }
+
+  // Creates a new branch instruction to |label_id|.
+  // Note that the user must makes sure the final basic block is
+  // well formed.
+  ir::Instruction* AddBranch(uint32_t label_id) {
+    std::unique_ptr<ir::Instruction> new_branch(new ir::Instruction(
+        GetContext(), SpvOpBranch, 0, 0,
+        {{spv_operand_type_t::SPV_OPERAND_TYPE_ID, {label_id}}}));
+    return AddInstruction(std::move(new_branch));
+  }
+
+  // Creates a new conditional instruction and the associated selection merge
+  // instruction if requested.
+  // The id |cond_id| is the id to the the condition instruction, must be of
+  // type bool.
+  // The id |true_id| is the basic block id to branch to if the condition is
+  // true.
+  // The id |false_id| is the basic block id to branch to if the condition is
+  // false.
+  // The id |merge_id| is the merge basic block id for the selection merge
+  // instruction. If |merge_id| equals kInvalidId then no selection merge
+  // instruction will be created.
+  // Note that the user must makes sure the final basic block is
+  // well formed.
+  ir::Instruction* AddBranchCond(uint32_t cond_id, uint32_t true_id,
```

Comment:
I'd rather not save the characters and just name this AddBranchConditional (or AddConditionalBranch).